docs/design.html

```diff
@@ -290,24 +290,58 @@ <h3 class="anchor-heading"><a id="semantics" class="anchor-link"></a><a href="#s
+    <p>
+    Kafka transactions are a bit different than transactions in other messaging systems. In Kafka, the consumer and producer are separate and it is only the producer which is transactional. It is however able to
+    make transactional updates to the consumer's position (confusingly called the "committed offset"), and it is this which gives the overall exactly-once behavior.
+    <p>
+    There are three key aspects to exactly-once processing using the producer and consumer, which match how Kafka Streams works.
+    <ol>
+        <li>The consumer uses partition assignment to ensure that it is the only consumer in the consumer group currently processing each partition.</li>
+        <li>The consumer uses read-committed isolation level to ensure that it does not consume records produced by transactions which aborted.</li>
+        <li>The producer uses transactions so that all of the records it produces, and any offsets it updates on behalf of the consumer, are performed atomically.</li>
+    </ol>
+    <p>
+    The consumer configuration must include <code>isolation.level=read_committed</code> and <code>enable.auto.commit=false</code>. The producer configuration must set <code>transactional.id</code>
+    to the name of the transactional ID to be used, which configures the producer for transactional delivery and also makes sure that a restarted application causes any in-flight transaction from
+    the previous instance to abort. Only the producer has the <code>transactional.id</code> configuration. Strictly speaking, the consumer doesn't have to use read_committed isolation level, but if
+    it does not, it will see records from aborted transactions and also open transactions which have not yet completed, which seems undesirable if trying to achieve exactly-once.
+    <p>
+    Here's an example of a <a href="https://github.com/apache/kafka/blob/trunk/tools/src/main/java/org/apache/kafka/tools/TransactionalMessageCopier.java">transactional message copier</a>
+    which uses these principles. It uses a <code>KafkaConsumer</code> to consume records from one topic and a <code>KafkaProducer</code> to produce records to another topic. It uses transactions to ensure
+    that there is no duplication or loss of records as they are copied.
+    <p>
+    It is important to handle exceptions and aborted transactions correctly. Any records written by the transational producer will be marked as being part of the transactions, and then when the
+    transaction commits or aborts, transaction marker records are written to indicate the outcome of the transaction. This is how the read-committed consumer does not see records from aborted
+    transactions. However, in the event of a transaction abort, the application's in-memory state and in particular the current position of the consumer must be reset explicitly so that it can
```

Comment:
```suggestion
    transactions. However, in the event of a transaction abort, the application's state and in particular the current position of the consumer must be reset explicitly so that it can
```